### ISSUE TITLE
[FR] Cleanup and finalize Attention Couple support

### ISSUE THREAD

**Issue by asagi4** [asagi4/comfyui-prompt-control#108]
See #107 

It seems to work, but needs a bit of cleanup.

**Comment by henrikvilhelmberglund** [comment]
The ppm version seems to have some issues with certain v-pred models: https://github.com/pamparamm/ComfyUI-ppm/issues/27
so if possible I would like to request support for both methods until it's fixed (if it's even fixable), something like
ppm: `ATTN()`
cgem156: `ATTN2()`

**Comment by asagi4** [comment]
I switched over to the ppm code with a couple tweaks, and it seems to work. I still need to figure out how to keep compatibility with negpip though.

**Comment by asagi4** [comment]
Seems like the outputs are completely different from ppm latent couple vs. mine. I wonder what broke, since I don't think I changed any of the actual calculations.

**Comment by pamparamm** [comment]
Interesting, for me the images are almost the same, but generation speed is much slower with prompt-control (perhaps due to how AND works?)

![Image](https://github.com/user-attachments/assets/cbc0f560-e13a-45a7-9330-ca02a501f7d7)

![Image](https://github.com/user-attachments/assets/d5421e9f-f59e-43be-bda4-5aa310dba118)

**Comment by asagi4** [comment]
But if you're getting nearly identical results consistently, then at least it's probably not the code that's broken, but something specific to my setup.

**Comment by pamparamm** [comment]
I believe 42cdfa0f5a213d1d827caee26ea21d301a52ef11 has broke FILL() behavior - now it produces intersecting masks:
**Before:**
`region1 FILL() ATTN() AND region2 MASK(0.5 1) ATTN()`:
![Image](https://github.com/user-attachments/assets/1e3f9414-47cf-44b1-94f3-c4be839671da)
**After:**
`region1 FILL() ATTN() AND region2 MASK(0.5 1) ATTN()`:
![Image](https://github.com/user-attachments/assets/5d5bcefc-44e9-4ade-89c6-0dbbad3707cc)
`region1 MASK(0 0.5) ATTN() AND region2 MASK(0.5 1) ATTN()`:
![Image](https://github.com/user-attachments/assets/743beee3-00d1-4c7d-90f7-966e24039ee9)

**Comment by pamparamm** [comment]
As for speed differences: the hook is causing conds and unconds to be processed in different batches.

Speaking of batches, there are some problems with `batch_size` > 1 on latest prompt-control version:

![Image](https://github.com/user-attachments/assets/4f2d2163-3d99-4585-a2a0-7c4aa5ff2d65)

**Comment by pamparamm** [comment]
I've found an edge case in ppm upstream related to uncond-only processing, TLDR:
replace
```python
        cond_output = torch.stack(cond_outputs).sum(0)
        outputs.append(cond_output)
```
with
```python
        if len(cond_outputs) > 0:
            cond_output = torch.stack(cond_outputs).sum(0)
            outputs.append(cond_output)
```

**Comment by asagi4** [comment]
Still not quite done, it seems at some point I broke longer prompts even though I thought I was testing them.

It's probably in the attn2 function since I pretty much just removed the loop and assumed the shapes worked out (which they did in testing)

**Comment by asagi4** [comment]
I hope the actual implementation is stable now. I think I might change the syntax a bit still, and maybe expose it as a hook node.
